File: 2020/Day 3/JavaScript/Solution.js
const fs = require('fs');

function PartOne(array) {
    var results = 0
    var position = 0

    array.forEach(element => {
        var currentChar = element.charAt(position)

        if (currentChar == "#") {
            results++
        }

        var positionToGo = position + 3
        if (positionToGo >= 31) {
            position = positionToGo - 31
        } else {
            position = position + 3
        }
    })
    return results
}

function PartTwo(array) {
    var finalResults = 1
    var found = []
    var numToCheck = ['1','3','5','7','1']
    var iter = 0

    numToCheck.forEach(nTC => {
        var results = 0
        var position = 0
        var rowCount = 0

        array.forEach(element => {
            if((iter == 4) && (!(rowCount % 2 == 0))){
                rowCount++
                return
            }

            var currentChar = element.charAt(position)

            if(currentChar == "#") { 
                results++
            }


            var positionToGo = parseInt(position) + parseInt(nTC)

            if (positionToGo >= 31) { 
                position = parseInt(positionToGo) - 31
            } else { 
                position = parseInt(position) + parseInt(nTC)
            }
            rowCount++
        })
        iter++
        found.push(results)
    })

    found.forEach(num => {
        finalResults *= num
    })

    return finalResults
}


fs.readFile('..\\Input.txt', 'UTF-8', (err, data) => {
    if (err) {
        console.log(err)
        return
    }

    var array = data.split('\n')
    results = PartOne(array)
    results2 = PartTwo(array)

    console.log(`Part One: ${results} Part Two: ${results2}`)
})

File: 2020/Day 3/Python/Solution.py
import re

with open('..\\input.txt') as file:
    lines = file.readlines()

def PartOne():
    position = 0
    results = 0

    for ele in lines:
        currentChar = ele[position]

        if currentChar == '#':
            results += 1

        positionToGo = position + 3
        
        if positionToGo >= 31:
            position = positionToGo - 31
        else:
            position = position + 3

    print("PartOne: " + str(results))


def PartTwo():
    finalResults = 1
    found = []
    numToCheck = [1,3,5,7,1]
    inter = 0

    for nTC in numToCheck:
        results = 0
        position = 0
        rowCount = 0

        for element in lines:
            if inter == 4 and (not(rowCount % 2 == 0)):
                rowCount = rowCount + 1
                continue
            
            currentChar = element[position]
            
            if currentChar == '#':
                results = results + 1

            positionToGo = int(position) + int(nTC)

            if positionToGo >= 31:
                position = int(positionToGo) - 31
            else:
                position = int(position) + int(nTC)

            rowCount = rowCount + 1

        inter = inter + 1
        found.append(results)

    for f in found:
        finalResults *= f

    print("PartTwo: " + str(finalResults))

PartOne()
PartTwo()